File: example-servers/go/services/huggingFace.go
package services

import (
	"encoding/json"
	"net/http"
	"errors"
	"bytes"
	"fmt"
	"io"
	"os"
)

// Make sure to set the HUGGING_FACE_API_KEY environment variable in a .env file (create if does not exist) - see .env.example

func HuggingFaceConversation(w http.ResponseWriter, r *http.Request) error {
	err, skipPreflight := ProcessIncomingRequest(&w, r)
	if err != nil { return err }
	if skipPreflight { return nil }

	conversationBodyBytes, err := createConversationRequestBodyBytes(w, r)
	if err != nil { return err }

	req, _ := http.NewRequest(
		"POST", "https://api-inference.huggingface.co/models/facebook/blenderbot-400M-distill", bytes.NewBuffer(conversationBodyBytes))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer " + os.Getenv("HUGGING_FACE_API_KEY"))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil { return err }
	
	defer resp.Body.Close()

	var resultData HuggingFaceConversationResult
	err = json.NewDecoder(resp.Body).Decode(&resultData)
	if err != nil { return err }

	if resultData.Error != "" {
    return errors.New(resultData.Error)
	}

	// Create response
	jsonResponse, err := CreateTextResponse(w, resultData.GeneratedText)
	if err != nil { return err }

	// Send response
	w.Header().Set("Content-Type", "application/json")
	w.Write(jsonResponse)
	return nil
}

func createConversationRequestBodyBytes(w http.ResponseWriter, r *http.Request) ([]byte, error) {
	// Text messages are stored inside request body using the Deep Chat JSON format:
	// https://deepchat.dev/docs/connect
	userRequestBodyBytes, err := io.ReadAll(r.Body)
	if err != nil { return userRequestBodyBytes, err }

	var deepChatRequestBody DeepChatRequestBody
	err = json.Unmarshal(userRequestBodyBytes, &deepChatRequestBody)
	if err != nil { return userRequestBodyBytes, err }

	conversationBody := createConversationBody(deepChatRequestBody.Messages)
	fmt.Println(conversationBody)
	return json.Marshal(conversationBody)
}

func createConversationBody(messages []DeepChatRequestMessage) HuggingFaceConversationBody {
	previousMessages := messages[:len(messages)-1]
	var pastUserInputs []string
	var generatedResponses []string
	for _, message := range previousMessages {
		if message.Role == "user" {
			pastUserInputs = append(pastUserInputs, message.Text)
		} else if message.Role == "ai" {
			generatedResponses = append(generatedResponses, message.Text)
		}
	}
	return HuggingFaceConversationBody{
		Inputs: HuggingFaceConversationInputs{
			PastUserInputs: pastUserInputs,
			GeneratedResponses: generatedResponses,
			Text: messages[len(messages)-1].Text,
		},
		WaitForModel: true,
	}
}

// You can use an example image here: https://github.com/OvidijusParsiunas/deep-chat/blob/main/example-servers/ui/assets/example-image.png
func HuggingFaceImage(w http.ResponseWriter, r *http.Request) error {
	err, skipPreflight := ProcessIncomingRequest(&w, r)
	if err != nil { return err }
	if skipPreflight { return nil }

	// Files are stored inside a form using Deep Chat request FormData format:
	// https://deepchat.dev/docs/connect
	err = r.ParseMultipartForm(32 << 20) // 32MB maximum file size
	if err != nil { return err }
	files := r.MultipartForm.File["files"]

	if (len(files) == 0) { return errors.New("no file was attached") }
	file, err := files[0].Open()
	if err != nil { return err }

	defer file.Close()

	req, err := http.NewRequest("POST", "https://api-inference.huggingface.co/models/google/vit-base-patch16-224", file)
	if err != nil { return err }

	req.Header.Set("Authorization", "Bearer " + os.Getenv("HUGGING_FACE_API_KEY"))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil { return err }

	defer resp.Body.Close()

	// Read the response body
	responseBody, err := io.ReadAll(resp.Body)
	if err != nil { return err }

	rawResponse := json.RawMessage(responseBody)

	// Check if the response is an error
	var errorData HugginFaceFileError
	if err := json.Unmarshal(rawResponse, &errorData); err == nil && errorData.Error != "" {
		return errors.New(errorData.Error)
	}

	// Check if the response is a successful result
	var resultData []HugginFaceImageResultLabel
	if err := json.Unmarshal(rawResponse, &resultData); err == nil {
		// Create response
		jsonResponse, err := CreateTextResponse(w, resultData[0].Label)
		if err != nil { return err }

		// Send response
		w.Header().Set("Content-Type", "application/json")
		w.Write(jsonResponse)
		return nil
	}

	return errors.New("unexpected response format")
}

// You can use an example audio here: https://github.com/OvidijusParsiunas/deep-chat/blob/main/example-servers/ui/assets/example-audio.m4a
func HuggingFaceSpeech(w http.ResponseWriter, r *http.Request) error {
	err, skipPreflight := ProcessIncomingRequest(&w, r)
	if err != nil { return err }
	if skipPreflight { return nil }

	// Files are stored inside a form using Deep Chat request FormData format:
	// https://deepchat.dev/docs/connect
	err = r.ParseMultipartForm(32 << 20) // 32MB maximum file size
	if err != nil { return err }
	files := r.MultipartForm.File["files"]

	if (len(files) == 0) { return errors.New("no file was attached") }
	file, err := files[0].Open()
	if err != nil { return err }

	defer file.Close()

	req, err := http.NewRequest("POST", "https://api-inference.huggingface.co/models/facebook/wav2vec2-large-960h-lv60-self", file)
	if err != nil { return err }

	req.Header.Set("Authorization", "Bearer " + os.Getenv("HUGGING_FACE_API_KEY"))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil { return err }

	defer resp.Body.Close()

	// Read the response body
	responseBody, err := io.ReadAll(resp.Body)
	if err != nil { return err }

	rawResponse := json.RawMessage(responseBody)

	// Check if the response is an error
	var errorData HugginFaceFileError
	if err := json.Unmarshal(rawResponse, &errorData); err == nil && errorData.Error != "" {
		return errors.New(errorData.Error)
	}

	// Check if the response is a successful result
	var resultData HugginFaceSpeechResult
	if err := json.Unmarshal(rawResponse, &resultData); err == nil {
		// Create response
		jsonResponse, err := CreateTextResponse(w, resultData.Text)
		if err != nil { return err }

		// Send response
		w.Header().Set("Content-Type", "application/json")
		w.Write(jsonResponse)
		return nil
	}

	return errors.New("unexpected response format")
}

File: example-servers/node/nestjs/src/services/huggingFace.ts
import {Injectable} from '@nestjs/common';
import {Request} from 'express';
import axios from 'axios';

// Make sure to set the HUGGING_FACE_API_KEY environment variable in a .env file (create if does not exist) - see .env.example

@Injectable()
export class HuggingFace {
  async conversation(body: Request['body']) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    const conversationBody = HuggingFace.createConversationBody(body.messages);
    const response = await axios.post(
      'https://api-inference.huggingface.co/models/facebook/blenderbot-400M-distill',
      conversationBody,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: 'Bearer ' + process.env.HUGGING_FACE_API_KEY,
        },
      }
    );
    // Sends response back to Deep Chat using the Response format:
    // https://deepchat.dev/docs/connect/#Response
    return {text: response.data.generated_text};
  }

  private static createConversationBody(messages: {text: string; role: string}[]) {
    const {text} = messages[messages.length - 1];
    const previousMessages = messages.slice(0, messages.length - 1);
    if (!text) return;
    const past_user_inputs = previousMessages.filter((message) => message.role === 'user').map((message) => message.text);
    const generated_responses = previousMessages.filter((message) => message.role === 'ai').map((message) => message.text);
    return {inputs: {past_user_inputs, generated_responses, text}, wait_for_model: true};
  }

  // You can use an example image here: https://github.com/OvidijusParsiunas/deep-chat/blob/main/example-servers/ui/assets/example-image.png
  async imageClassification(files: Array<Express.Multer.File>) {
    const parseResult = (result: any) => result[0].label;
    const url = 'https://api-inference.huggingface.co/models/google/vit-base-patch16-224';
    return this.sendFile(files, url, parseResult);
  }

  // You can use an example audio file here: https://github.com/OvidijusParsiunas/deep-chat/blob/main/example-servers/ui/assets/example-audio.m4a
  async speechRecognition(files: Array<Express.Multer.File>) {
    const parseResult = (result: any) => result.text;
    const url = 'https://api-inference.huggingface.co/models/facebook/wav2vec2-large-960h-lv60-self';
    return this.sendFile(files, url, parseResult);
  }

  private async sendFile(files: Array<Express.Multer.File>, url: string, parseResult: (result: any) => string) {
    // Files are stored inside a form using Deep Chat request FormData format:
    // https://deepchat.dev/docs/connect
    const response = await axios.post(url, files[0].buffer, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: 'Bearer ' + process.env.HUGGING_FACE_API_KEY,
      },
    });
    // Sends response back to Deep Chat using the Response format:
    // https://deepchat.dev/docs/connect/#Response
    return {text: parseResult(response.data)};
  }
}
